docs/remediation-controller.md

```diff
@@ -0,0 +1,136 @@
+
+```yaml
+
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  name: worker-healthcheck
+  namespace: metal3
+spec:
+  # clusterName is required to associate this MachineHealthCheck with a particular cluster
+  clusterName: test1
+  # (Optional) maxUnhealthy prevents further remediation if the cluster is already partially unhealthy
+  maxUnhealthy: 100%
+  # (Optional) nodeStartupTimeout determines how long a MachineHealthCheck should wait for
+  # a Node to join the cluster, before considering a Machine unhealthy.
+  # Defaults to 10 minutes if not specified.
+  # Set to 0 to disable the node startup timeout.
+  # Disabling this timeout will prevent a Machine from being considered unhealthy when
+  # the Node it created has not yet registered with the cluster. This can be useful when
+  # Nodes take a long time to start up or when you only want condition based checks for
+  # Machine health.
+  nodeStartupTimeout: 0m
+  # selector is used to determine which Machines should be health checked
+  selector:
+    matchLabels:
+      nodepool: nodepool-0
+  # Conditions to check on Nodes for matched Machines, if any condition is matched for the duration of its timeout, the Machine is considered unhealthy
+  unhealthyConditions:
+  - type: Ready
+    status: Unknown
+    timeout: 300s
+  - type: Ready
+    status: "False"
+    timeout: 300s
+  remediationTemplate: # added infrastructure reference
+    kind: Metal3RemediationTemplate
+    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+    name: worker-remediation-request
+
+```
+
+```yaml
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3RemediationTemplate
+metadata:
+    name: worker-remediation-request
+    namespace: metal3
+spec:
+  template:
+    spec:
+      strategy:
+        type: "Reboot"
+        retryLimit: 2
+        timeout: 300s
+
+```
+
+Use the following examples as a basis for creating a MachineHealthCheck and Metal3Remediation for controlplane nodes:
+
+```yaml
+
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  name: controlplane-healthcheck
+  namespace: metal3
+spec:
+  clusterName: test1
+  maxUnhealthy: 100%
+  nodeStartupTimeout: 0m
+  selector:
+    matchLabels:
+      cluster.x-k8s.io/control-plane: ""
+  unhealthyConditions:
+    - type: Ready
+      status: Unknown
+      timeout: 300s
+    - type: Ready
+      status: "False"
+      timeout: 300s
+  remediationTemplate: # added infrastructure reference
+    kind: Metal3RemediationTemplate
+    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+    name: controlplane-remediation-request
+
+```
+
+```yaml
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3RemediationTemplate
+metadata:
+    name: controlplane-remediation-request
+    namespace: metal3
+spec:
+  template:
+    spec:
+      strategy:
+        type: "Reboot"
+        retryLimit: 1
```

Comment:
I've noticed from the code that if `retryLimit` is not set, or if it's omitted, no reboot will be done and instead we skip to Machine deletion.

Reply:
Is that desirable @macaptain or shall we give some default values here and make sure that reboot is tried atleast once.

Reply:
I don't think it's necessarily undesirable, but we should add a comment in the documentation and in the API at least.

Reply:
Allright. Lets take this in and then you can add a follow up PR with the changes.